File: deepseek_client.py
"""
DeepSeek API客户端模块
用于调用DeepSeek模型进行文本分析和数据提取
"""

import os
import json
import requests
from typing import Dict, List, Any, Optional
from dotenv import load_dotenv

# 加载环境变量
load_dotenv()

class DeepSeekClient:
    """DeepSeek API客户端"""

    def __init__(self, api_key: Optional[str] = None):
        """
        初始化DeepSeek客户端

        Args:
            api_key: DeepSeek API密钥，如果不提供则从环境变量获取
        """
        self.api_key = api_key or os.getenv('DEEPSEEK_API_KEY')
        if not self.api_key:
            raise ValueError("请设置DEEPSEEK_API_KEY环境变量或直接传入api_key参数")

        self.base_url = "https://api.deepseek.com/v1/chat/completions"
        self.headers = {
            "Authorization": f"Bearer {self.api_key}",
            "Content-Type": "application/json"
        }

    def extract_data_from_text(self, text: str, extraction_type: str = "comprehensive") -> Dict[str, Any]:
        """
        从文本中提取数据和重要信息

        Args:
            text: 要分析的文本
            extraction_type: 提取类型 ("comprehensive", "entities", "sentiment", "keywords")

        Returns:
            提取的数据字典
        """
        prompts = {
            "comprehensive": """
            请分析以下文本，提取以下信息并以JSON格式返回：
            1. 关键实体（人物、地点、组织、产品等）
            2. 重要数字和统计数据
            3. 时间信息
            4. 情感倾向（正面/负面/中性）
            5. 主要主题和关键词
            6. 重要事件或趋势
            7. 可量化的指标

            请确保返回的数据结构清晰，便于后续可视化处理。

            文本内容：
            """,
            "entities": """
            请从以下文本中提取所有重要实体，包括：
            - 人名
            - 地名
            - 组织名
            - 产品名
            - 其他重要名词

            以JSON格式返回，包含实体类型和实体名称。

            文本内容：
            """,
            "sentiment": """
            请分析以下文本的情感倾向，包括：
            - 整体情感（正面/负面/中性）
            - 情感强度（1-10分）
            - 关键情感词汇
            - 情感变化趋势

            以JSON格式返回分析结果。

            文本内容：
            """,
            "keywords": """
            请从以下文本中提取关键词，包括：
            - 高频词汇
            - 重要概念
            - 专业术语
            - 主题词

            以JSON格式返回，包含词汇和频率。

            文本内容：
            """
        }

        prompt = prompts.get(extraction_type, prompts["comprehensive"]) + text

        try:
            response = self._make_request(prompt)
            return self._parse_response(response)
        except Exception as e:
            print(f"数据提取失败: {e}")
            return {"error": str(e)}

    def generate_visualization_suggestions(self, extracted_data: Dict[str, Any]) -> List[Dict[str, Any]]:
        """
        基于提取的数据生成可视化建议

        Args:
            extracted_data: 提取的数据

        Returns:
            可视化建议列表
        """
        prompt = f"""
        基于以下提取的数据，请建议最适合的可视化图表类型和配置：

        数据内容：
        {json.dumps(extracted_data, ensure_ascii=False, indent=2)}

        请为每种数据类型推荐：
        1. 图表类型（柱状图、折线图、饼图、散点图、词云等）
        2. 图表配置参数
        3. 数据映射方式
        4. 颜色方案建议
        5. 交互功能建议

        以JSON格式返回建议列表。
        """

        try:
            response = self._make_request(prompt)
            return self._parse_response(response)
        except Exception as e:
            print(f"可视化建议生成失败: {e}")
            return [{"error": str(e)}]

    def _make_request(self, prompt: str) -> Dict[str, Any]:
        """
        向DeepSeek API发送请求

        Args:
            prompt: 提示词

        Returns:
            API响应
        """
        payload = {
            "model": "deepseek-chat",
            "messages": [
                {
                    "role": "user",
                    "content": prompt
                }
            ],
            "temperature": 0.3,
            "max_tokens": 4000
        }

        response = requests.post(
            self.base_url,
            headers=self.headers,
            json=payload,
            timeout=30
        )

        if response.status_code != 200:
            raise Exception(f"API请求失败: {response.status_code} - {response.text}")

        return response.json()

    def _parse_response(self, response: Dict[str, Any]) -> Dict[str, Any]:
        """
        解析API响应

        Args:
            response: API响应

        Returns:
            解析后的数据
        """
        try:
            content = response['choices'][0]['message']['content']
            # 尝试解析JSON
            if content.strip().startswith('{') or content.strip().startswith('['):
                return json.loads(content)
            else:
                # 如果不是JSON格式，返回原始内容
                return {"content": content}
        except (KeyError, json.JSONDecodeError) as e:
            return {"error": f"响应解析失败: {e}", "raw_response": response}

File: src/deepseek-client.ts
/**
 * DeepSeek API客户端模块
 * 用于调用DeepSeek模型进行文本分析和数据提取
 */

import axios, { AxiosResponse } from 'axios';
import * as dotenv from 'dotenv';
import { DeepSeekResponse, ExtractedData, VisualizationSuggestion, DeepSeekClientConfig } from './types';

// 加载环境变量
dotenv.config();

export class DeepSeekClient {
  private apiKey: string;
  private baseUrl: string;
  private timeout: number;

  constructor(config?: DeepSeekClientConfig) {
    this.apiKey = config?.apiKey || process.env.DEEPSEEK_API_KEY || '';
    if (!this.apiKey) {
      throw new Error('请设置DEEPSEEK_API_KEY环境变量或直接传入apiKey参数');
    }

    this.baseUrl = config?.baseUrl || 'https://api.deepseek.com/chat/completions';
    this.timeout = config?.timeout || 120000;
  }

  /**
   * 从文本中提取数据和重要信息
   * @param text 要分析的文本
   * @returns 提取的数据字典
   */
  async extractDataFromText(text: string): Promise<ExtractedData> {
    const prompt = `
分析以下文本，提取其中的数据和重要信息

提取要求：
1. 数据焦点：重点提取可以量化或对比的数据点。
2. 关系明确：清晰标注数据点之间的关系（例如：分类、时间序列、占比等）。
3. 信息来源：对于每个数据点，简要说明它在文本中代表的意义。

请直接以结构化的JSON格式返回，确保信息足够清晰且便于后续处理。

示例如下：
1. 如果段落中有指标数值相关，如下：
国家外汇管理局9月7日公布的数据显示，截至2025年8月末，我国外汇储备规模为33221.54亿美元，较7月末增加299.19亿美元，升幅为0.91%。
则返回：
\`\`\`json
{
  "event_summary": "中国外汇储备规模在2025年8月末的变化。",
  "source": "国家外汇管理局",
  "data_as_of": "09-07",
  "data_points": [
    {
      "period_end": "2025年8月末",
      "value_usd_billion": 33221.54,
      "description": "期末外汇储备总规模"
    },
    {
      "period_end": "2025年7月末",
      "value_usd_billion": 32922.35,
      "description": "上期外汇储备总规模（33221.54 - 299.19）"
    }
  ]
}
\`\`\`

2. 如果段落中没有指标数值相关，如下：
国家外汇局表示，2025年8月，受主要经济体货币政策预期、宏观经济数据等因素影响，美元指数下跌，全球金融资产价格总体上涨。汇率折算和资产价格变化等因素综合作用，当月外汇储备规模上升。
则返回：
\`\`\`json
{
  "event_summary": "国家外汇储备规模上升",
  "source": "国家外汇局",
  "data_as_of": "09-07",
  "event_list": [
    {
      "event_time": "2025年8月",
      "metric_name": "美元指数",
      "action": "下跌",
      "category": "市场表现",
    },
    {
      "event_time": "2025年8月",
      "metric_name": "全球金融资产价格",
      "action": "总体上涨",
      "category": "市场表现",
    },
    {
      "event_time": "2025年8月",
      "metric_name": "外汇储备规模",
      "action": "上升",

    }
  ]
}
\`\`\`

文本内容：
${text}`;

    try {
      const response = await this.makeRequest(prompt);
      return this.parseResponse(response);
    } catch (error) {
      console.error('数据提取失败:', error);
      throw error;
    }
  }

  /**
   * 基于提取的数据生成可视化建议
   * @param extractedData 提取的数据
   * @returns 可视化建议列表
   */
  generateVisualizationSuggestions(): VisualizationSuggestion[] {
    // TODO: 实现可视化建议生成逻辑
    return [];
  }

  /**
   * 向DeepSeek API发送请求
   * @param prompt 提示词
   * @returns API响应
   */
  private async makeRequest(prompt: string): Promise<DeepSeekResponse> {
    const payload = {
      model: 'deepseek-chat',
      messages: [
        {
          role: 'user',
          content: prompt,
        },
      ],
      temperature: 0.7,
      max_tokens: 4000,
    };

    try {
      const response: AxiosResponse<DeepSeekResponse> = await axios.post(this.baseUrl, payload, {
        headers: {
          Authorization: `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
        },
        timeout: this.timeout,
      });

      if (response.status !== 200) {
        throw new Error(`API请求失败: ${response.status} - ${JSON.stringify(response.data)}`);
      }

      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error)) {
        throw new Error(`API请求失败: ${error.response?.status} - ${JSON.stringify(error.response?.data)}`);
      }
      throw error;
    }
  }

  /**
   * 解析API响应
   * @param response API响应
   * @returns 解析后的数据
   */
  private parseResponse(response: DeepSeekResponse): ExtractedData {
    try {
      // 提取AI生成的内容
      const content = response.choices[0].message.content;

      // 尝试解析JSON格式的内容
      try {
        return JSON.parse(content);
      } catch (jsonError) {
        // 如果直接解析失败，尝试提取markdown代码块中的JSON
        const jsonContent = this.extractJsonFromMarkdown(content);
        if (jsonContent) {
          try {
            return JSON.parse(jsonContent);
          } catch (jsonParseError) {
            // 如果仍然解析失败，返回原始内容
            return { content, format: 'text' } as any;
          }
        }

        // 如果都不是有效的JSON，返回原始内容
        return { content, format: 'text' } as any;
      }
    } catch (error) {
      throw new Error(`响应解析失败: ${error}`);
    }
  }

  /**
   * 从markdown代码块中提取JSON内容
   * @param content 包含markdown代码块的字符串
   * @returns 提取的JSON字符串，如果没有找到则返回null
   */
  private extractJsonFromMarkdown(content: string): string | null {
    // 匹配 ```json ... ``` 格式的代码块
    const jsonPattern = /```json\s*\n(.*?)\n```/s;
    const jsonMatch = content.match(jsonPattern);

    if (jsonMatch) {
      return jsonMatch[1].trim();
    }

    // 如果没有找到json标记的代码块，尝试匹配任何代码块
    const codePattern = /```\s*\n(.*?)\n```/s;
    const codeMatch = content.match(codePattern);

    if (codeMatch) {
      return codeMatch[1].trim();
    }

    return null;
  }
}
